src/message_bus.zig

```diff
@@ -73,60 +191,413 @@ pub const MessageBus = struct {
+                self.recv_body();
+                return;
+            }
+            assert(self.bytes_received == self.incoming_message.buffer.len);
+
+            const data = self.incoming_message.buffer[@sizeOf(Header)..self.incoming_header.size];
+            if (self.incoming_header.valid_checksum_data(data)) {
+                self.message_bus.server.on_message(self.incoming_message);
+            } else {
+                log.err("invalid checksum on data received from {}", .{self.peer});
+                self.close();
+            }
+
+            self.incoming_message.references -= 1;
+            self.message_bus.gc(self.incoming_message);
+
+            // Reset state and try to receive the next message.
+            self.incoming_header = undefined;
+            self.incoming_message = undefined;
+            self.bytes_received = 0;
+            self.recv_header();
+        }
+
+        fn send_message(self: *Connection, envelope: *Envelope) void {
+            assert(self.peer == .client);
+            const queue_was_empty = self.send_queue.out == null;
+            self.send_queue.push(envelope);
+            // If the queue was not empty, the message will be sent after the
+            // messages currently being sent.
+            if (queue_was_empty) self.send();
+        }
+
+        fn send(self: *ReplicaConnection) void {
+            // If currently disconnected, do nothing.
+            // This function will be called again on reconnect.
+            if (self.fd == -1) return;
+            const envelope = self.send_queue.out orelse return;
+            self.message_bus.io.send(
+                *ReplicaConnection,
+                self,
+                complete_send,
+                &self.completion,
+                self.fd,
+                envelope.message.buffer[self.bytes_sent..],
+                os.MSG_NOSIGNAL,
+            );
+        }
+
+        fn complete_send(self: *ReplicaConnection, completion: *IO.Completion, result: SendError!usize) void {
+            self.bytes_sent += result catch |err| {
+                // TODO: maybe don't need to close on *every* error
+                log.err("error sending message to replica at {}: {}", .{ self.address, err });
+                self.close();
+                return;
+            };
+            assert(self.bytes_sent <= self.send_queue.out.?.message.buffer.len);
+            // If the message has been fully sent, move on to the next one.
+            if (self.bytes_sent == self.send_queue.out.?.message.buffer.len) {
+                self.bytes_sent = 0;
+                const envelope = self.send_queue.pop().?;
+                envelope.message.references -= 1;
+                self.message_bus.gc(envelope.message);
+                self.message_bus.allocator.destroy(envelope);
+            }
+            self.send();
+        }
+
+        fn close(self: *Connection) void {
+            // If an error occurs in both sending and receving at roughly,
```

Comment:
Great catch!

Reply:
It would be if I had remembered to set `self.fd = -1` at the end of this function. Fixed now though :)